File: 0407-trapping-rain-water-ii/0407-trapping-rain-water-ii.java
import java.util.PriorityQueue;

class Solution {

    public int trapRainWater(int[][] heightMap) {
        if (heightMap == null || heightMap.length == 0 || heightMap[0].length == 0) {
            return 0;
        }

        int m = heightMap.length, n = heightMap[0].length;
        boolean[][] visited = new boolean[m][n];
        PriorityQueue<Cell> pq = new PriorityQueue<>((a, b) -> a.height - b.height);

        // Add all boundary cells to the priority queue
        for (int i = 0; i < m; i++) {
            pq.offer(new Cell(i, 0, heightMap[i][0]));
            pq.offer(new Cell(i, n - 1, heightMap[i][n - 1]));
            visited[i][0] = true;
            visited[i][n - 1] = true;
        }
        for (int j = 0; j < n; j++) {
            pq.offer(new Cell(0, j, heightMap[0][j]));
            pq.offer(new Cell(m - 1, j, heightMap[m - 1][j]));
            visited[0][j] = true;
            visited[m - 1][j] = true;
        }

        int totalWater = 0;
        int[] dirX = {-1, 1, 0, 0};
        int[] dirY = {0, 0, -1, 1};

        // Process cells in priority queue
        while (!pq.isEmpty()) {
            Cell cell = pq.poll();

            for (int k = 0; k < 4; k++) {
                int x = cell.x + dirX[k];
                int y = cell.y + dirY[k];

                if (x >= 0 && x < m && y >= 0 && y < n && !visited[x][y]) {
                    visited[x][y] = true;
                    // Water trapped depends on the difference between the current height and neighbor height
                    totalWater += Math.max(0, cell.height - heightMap[x][y]);
                    // Update the height of the neighbor to max(cell height, neighbor height)
                    pq.offer(new Cell(x, y, Math.max(cell.height, heightMap[x][y])));
                }
            }
        }

        return totalWater;
    }

    // Helper class for the priority queue
    static class Cell {
        int x, y, height;

        Cell(int x, int y, int height) {
            this.x = x;
            this.y = y;
            this.height = height;
        }
    }
}

File: 0407-trapping-rain-water-ii/0407-trapping-rain-water-ii.py
class Solution:
    def trapRainWater(self, heightMap: List[List[int]]) -> int:
        import heapq
        if not heightMap or not heightMap[0]:
            return 0

        m, n = len(heightMap), len(heightMap[0])

        # Min-Heap Priority Queue: stores (height, x, y)
        min_heap = []
        visited = set()

        # Add all the boundary cells to the heap
        for i in range(m):
            for j in range(n):
                if i == 0 or i == m - 1 or j == 0 or j == n - 1:
                    heapq.heappush(min_heap, (heightMap[i][j], i, j))
                    visited.add((i, j))

        # Directions for moving up, down, left, right
        directions = [(-1, 0), (1, 0), (0, -1), (0, 1)]
        water_trapped = 0

        # Process the heap
        while min_heap:
            height, x, y = heapq.heappop(min_heap)

            # Explore the four neighbors
            for dx, dy in directions:
                nx, ny = x + dx, y + dy

                # Check if the neighbor is within bounds and not visited
                if 0 <= nx < m and 0 <= ny < n and (nx, ny) not in visited:
                    # Calculate trapped water
                    if heightMap[nx][ny] < height:
                        water_trapped += height - heightMap[nx][ny]

                    # Add the neighbor to the heap with the new height (max of current height and neighbor height)
                    heapq.heappush(min_heap, (max(height, heightMap[nx][ny]), nx, ny))
                    visited.add((nx, ny))

        return water_trapped

        